p2p/src/main/java/bisq/network/p2p/network/Connection.java

```diff
@@ -882,4 +864,51 @@ public void run() {
             handleException(t);
         }
     }
+
+    protected boolean handleSupportedCapabilitiesMessage(NetworkEnvelope networkEnvelope) {
+        Capabilities supportedCapabilities = ((SupportedCapabilitiesMessage) networkEnvelope).getSupportedCapabilities();
+        if (supportedCapabilities == null || supportedCapabilities.isEmpty()) {
+            return false;
+        }
+
+        if (this.capabilities.equals(supportedCapabilities)) {
+            return false;
+        }
+
+        if (!Capabilities.hasMandatoryCapability(supportedCapabilities)) {
+            log.info("We close a connection because of " +
+                            "CloseConnectionReason.MANDATORY_CAPABILITIES_NOT_SUPPORTED " +
+                            "to node {}. Capabilities of old node: {}, " +
+                            "networkEnvelope class name={}",
+                    getSenderNodeAddressAsString(networkEnvelope),
+                    supportedCapabilities.prettyPrint(),
+                    networkEnvelope.getClass().getSimpleName());
+            shutDown(CloseConnectionReason.MANDATORY_CAPABILITIES_NOT_SUPPORTED);
+            return true;
+        }
+
+        this.capabilities.set(supportedCapabilities);
+
+        capabilitiesListeners.forEach(weakListener -> {
+            SupportedCapabilitiesListener supportedCapabilitiesListener = weakListener.get();
+            if (supportedCapabilitiesListener != null) {
+                UserThread.execute(() -> supportedCapabilitiesListener.onChanged(supportedCapabilities));
+            }
+        });
+        return false;
+    }
+
+    @Nullable
+    private NodeAddress getSenderNodeAddress(NetworkEnvelope networkEnvelope) {
+        return getPeersNodeAddressOptional().isPresent() ?
+                getPeersNodeAddressOptional().get() :
+                networkEnvelope instanceof SendersNodeAddressMessage ?
+                        ((SendersNodeAddressMessage) networkEnvelope).getSenderNodeAddress() :
+                        null;
```

Comment:
There is `orElse` for Optional
```suggestion
        return getPeersNodeAddressOptional().orElse(
                networkEnvelope instanceof SendersNodeAddressMessage ?
                        ((SendersNodeAddressMessage) networkEnvelope).getSenderNodeAddress() :
                        null);
```